### PR TITLE
[#9] [Backend] As a user, I can log in with email and password

### PR DESCRIPTION
**Pull request by Shayokh144**
#9 

## What happened 👀

- Add `LoginService` class to create a post request for getting login token
- Add `LoginRequest` class to create json request body
- Add `LoginResponse` class to decode json response
- Use jpax for decoding response
- Add `LoginModel` to assign login token data generated by a successful login event 
- Add `LoginRepositoryImpl` and `LoginUseCase` to provide login facility
- Add env variable for client secret and client id
- Add test for repository and use-case  in `login_repository_test.dart` and `login_use_case_test.dart`
- Add dummy login data for test

## Insight 📝

- Construct Json request body using client id, client secret
- Send login request, receive response
- Flatten and decode response to readable object using `jpax`
- Added unit test to check both login success and login failed cases

## Proof Of Work 📹

- All test case executed 
- Received login response from API for correct credentials
- Sample Request:

REQUEST

- Sample Response:

RESPONSE

### CODE REVIEW

File: lib/usecases/login_use_case.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import 'package:flutter_survey/model/login_model.dart';
+
+@Injectable()
+class LoginUseCase extends UseCase<LoginModel, LoginRequest> {
+  final LoginRepository _loginRepository;
+
+  const LoginUseCase(this._loginRepository);
+
+  @override
+  Future<Result<LoginModel>> call(LoginRequest input) {
```
Comment (Tuubz) [inline lib/usecases/login_use_case.dart:15]:
```suggestion
  Future<Result<LoginModel>> call(LoginInput input) {
```

Should we create a `LoginInput` class, to be more specific on what's being inputted from the UI layer?
```
LoginInput({
  required this.email,
  required this.password,
});
```

Inside `login_repository.dart`, we create the `LoginRequest` object, so it's specific to the network layer:

```
LoginRequest loginRequest = LoginRequest(
  grantType: "password",
  email: email,
  password: password,
  clientId: Env.restApiClientId,
  clientSecret: Env.restApiClientSecret,
);

final response = await _loginService.doLogin(request);
```

Comment (Shayokh144) [reply lib/usecases/login_use_case.dart:15]: @Tuubz I did exactly this, but it creates problem when I wrote test for usecase class. In this style `Env.restApiClientId` and `Env.restApiClientSecret` cannot be read when try to read from test method call.

Comment (Tuubz) [reply lib/usecases/login_use_case.dart:15]: @Shayokh144 Would this be able to resolve the issue? 🙇 

```
  FlutterConfig.loadValueForTesting({
    'CLIENT_ID': 'CLIENT_ID',
    'CLIENT_SECRET': 'CLIENT_SECRET',
  });
```

**Reference**: https://github.com/chornerman/survey-flutter/blob/develop/test/api/repository/auth_repository_test.dart

Comment (Shayokh144) [reply lib/usecases/login_use_case.dart]: @Tuubz thanks, it works. I updated the code.

**Review by Tuubz — APPROVED** [review]
💪 💪 💪

File: lib/api/service/login_service.dart [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import 'package:dio/dio.dart';
+import 'package:flutter_survey/api/request/login_request.dart';
+import 'package:flutter_survey/api/response/login_response.dart';
+import 'package:retrofit/retrofit.dart';
+
+part 'login_service.g.dart';
+
+abstract class LoginService {
+  Future<LoginResponse> doLogin(@Body() LoginRequest request);
```
Comment (luongvo) [inline lib/api/service/login_service.dart:9]: the better name could be `logIn` as a verb

File: lib/env.dart [hunk 3]
```diff
@@ -4,4 +4,12 @@ class Env {
...
     return FlutterConfig.get('REST_API_ENDPOINT');
   }
+
+  static String get restApiClientId {
+    return FlutterConfig.get('REST_API_CLIENT_ID');
+  }
+
+  static String get restApiClientSecret {
+    return FlutterConfig.get('REST_API_CLIENT_SECRET');
+  }
```
Comment (luongvo) [inline lib/env.dart:14]: Do you plan to update the `.env.sample` for instruction?

File: lib/ui/login/login_screen.dart [hunk 4]
```diff
@@ -173,6 +180,23 @@ class _LoginScreenState extends State<LoginScreen>
     );
   }
 
+  // TODO: remove this dummy method in [#10]
+  void doLogin() async {
```
Comment (luongvo) [inline lib/ui/login/login_screen.dart:184]: same, could be `logIn`

File: lib/api/repository/login_repository.dart [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import 'package:flutter_survey/api/service/login_service.dart';
+import 'package:flutter_survey/model/login_model.dart';
+import 'package:flutter_survey/api/exception/network_exceptions.dart';
+import 'package:flutter_survey/usecases/login_use_case.dart';
+import 'package:injectable/injectable.dart';
+import 'package:flutter_survey/env.dart';
+
+abstract class LoginRepository {
+  Future<LoginModel> doLogin({
+    required LoginInput input,
```
Comment (luongvo) [inline lib/api/repository/login_repository.dart:11]: I think the Repository layer should NOT receive the `input` model. it should receive separate `email` and `password` params instead 💭 The thing is passing a single model `LoginInput` could be so clear as the API parameters.

File: lib/api/repository/login_repository.dart [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import 'package:flutter_survey/api/request/login_request.dart';
+import 'package:flutter_survey/api/service/login_service.dart';
+import 'package:flutter_survey/model/login_model.dart';
+import 'package:flutter_survey/api/exception/network_exceptions.dart';
+import 'package:flutter_survey/usecases/login_use_case.dart';
+import 'package:injectable/injectable.dart';
+import 'package:flutter_survey/env.dart';
+
+abstract class LoginRepository {
+  Future<LoginModel> doLogin({
```
Comment (luongvo) [inline lib/api/repository/login_repository.dart:10]: the same could be `logIn`

File: lib/api/repository/login_repository.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    required LoginInput input,
+  });
+}
+
+@Singleton(as: LoginRepository)
+class LoginRepositoryImpl extends LoginRepository {
+  final LoginService _loginService;
+
+  LoginRepositoryImpl(this._loginService);
+  @override
```
Comment (luongvo) [inline lib/api/repository/login_repository.dart:20]: we need 1 line break here
```suggestion
  LoginRepositoryImpl(this._loginService);

  @override
```

File: lib/api/repository/login_repository.dart [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  Future<LoginModel> doLogin({
+    required LoginInput input,
+  }) async {
+    try {
+      final LoginRequest loginRequest = LoginRequest(
+        grantType: "password",
```
Comment (luongvo) [inline lib/api/repository/login_repository.dart:26]: please extract this to a const, maybe keep it in the same file `LoginRequest`

File: lib/api/repository/login_repository.dart [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import 'package:flutter_survey/api/request/login_request.dart';
+import 'package:flutter_survey/api/service/login_service.dart';
+import 'package:flutter_survey/model/login_model.dart';
+import 'package:flutter_survey/api/exception/network_exceptions.dart';
+import 'package:flutter_survey/usecases/login_use_case.dart';
+import 'package:injectable/injectable.dart';
+import 'package:flutter_survey/env.dart';
+
+abstract class LoginRepository {
```
Comment (luongvo) [inline lib/api/repository/login_repository.dart:9]: This repository will not be used for the login action only, right? Can we rename it to a common name, such as AuthenticationRepository?

File: lib/api/service/login_service.dart [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import 'package:dio/dio.dart';
+import 'package:flutter_survey/api/request/login_request.dart';
+import 'package:flutter_survey/api/response/login_response.dart';
+import 'package:retrofit/retrofit.dart';
+
+part 'login_service.g.dart';
+
+abstract class LoginService {
```
Comment (luongvo) [inline lib/api/service/login_service.dart:8]: Same, AuthenticationService?

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: lib/api/repository/authentication_repository.dart [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import 'package:flutter_survey/api/request/login_request.dart';
+import 'package:flutter_survey/api/service/authentication_service.dart';
+import 'package:flutter_survey/model/login_model.dart';
+import 'package:flutter_survey/api/exception/network_exceptions.dart';
+import 'package:injectable/injectable.dart';
+import 'package:flutter_survey/env.dart';
+
+abstract class AuthenticationRepository {
+  Future<LoginModel> login({
```
Comment (luongvo) [inline lib/api/repository/authentication_repository.dart:9]: `logIn` as a verb, not `login` is noun please 🙏

File: lib/usecases/login_use_case.dart [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  final String password;
+
+  LoginInput({
+    required this.email,
+    required this.password,
+  });
+}
+
+@Injectable()
+class LoginUseCase extends UseCase<LoginModel, LoginInput> {
```
Comment (luongvo) [inline lib/usecases/login_use_case.dart:18]: Optional: usecase should start with a verb too, while the model should be a noun

```suggestion
class LogInUseCase extends UseCase<LoginModel, LoginInput> {
```

File: lib/usecases/login_use_case.dart
Comment (luongvo) [inline lib/usecases/login_use_case.dart:1]: Then, the file name should be `log_in_use_case`

File: lib/ui/login/login_screen.dart [hunk 13]
```diff
@@ -182,6 +189,23 @@ class _LoginScreenState extends State<LoginScreen>
     );
   }
 
+  // TODO: remove this dummy method in [#10]
+  void login() async {
```
Comment (luongvo) [inline lib/ui/login/login_screen.dart:193]: Optinal: action should start with a verb too

```suggestion
  void logIn() async {
```

**Review by luongvo — APPROVED** [review]
Some minor issues, you can fix them in following PRs

**Comment by Tuubz** [comment]
@Shayokh144 Could you please rebase? 🙏

**Pull request merged**